core/src/Kokkos_View.hpp

```diff
@@ -75,25 +75,55 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
          (i6 != KOKKOS_INVALID_INDEX) + (i7 != KOKKOS_INVALID_INDEX);
 }
 
-KOKKOS_INLINE_FUNCTION
-void runtime_check_rank(const size_t rank, const size_t dyn_rank,
-                        const bool is_void_spec, const size_t i0,
-                        const size_t i1, const size_t i2, const size_t i3,
-                        const size_t i4, const size_t i5, const size_t i6,
-                        const size_t i7, const std::string& label) {
+// FIXME Ideally, we would not instantiate this function for every possible View
+// type. We should be able to only pass "extent" when we use mdspan.
+template <typename View>
+KOKKOS_INLINE_FUNCTION void runtime_check_rank(
+    const View&, const size_t rank, const size_t dyn_rank,
+    const bool is_void_spec, const size_t i0, const size_t i1, const size_t i2,
+    const size_t i3, const size_t i4, const size_t i5, const size_t i6,
+    const size_t i7, const char* label) {
   (void)(label);
 
   if (is_void_spec) {
     const size_t num_passed_args =
         count_valid_integers(i0, i1, i2, i3, i4, i5, i6, i7);
+    // We either allow to pass as many extents as the dynamic rank is, or
+    // as many extents as the total rank is. In the latter case, the given
+    // extents for the static dimensions must match the
+    // compile-time extents.
+    const bool n_args_is_dyn_rank = num_passed_args == dyn_rank;
+    const bool n_args_is_rank     = num_passed_args == rank;
+
+    if (n_args_is_rank) {
```

Comment:
maybe add a `if constexpr (View::rank()!=View::rank_dynamic())` around this?

Reply:
I think I still want the if constexpr around the whole thing. At a minimum you are counting the valid integers unnecessarily I think.

Reply:
I added the check in 90e8cc9c96c0918b627f9bd5e9ffcce6025dd6c9.